### PR TITLE
4103 enhances surface Dice to use subvoxel borders

### LINKED ISSUE #4103: extensions of the surface-based metric
**Is your feature request related to a problem? Please describe.**
Follow up of https://github.com/Project-MONAI/MONAI/pull/4050,
this is a feature request to extend the surface-based metric module for 3D, with the support of more detailed quantizations when computing local surface lengths/areas (such as the computation proposed in https://github.com/deepmind/surface-distance).

https://github.com/Project-MONAI/MONAI/pull/4050#discussion_r842152042
> the way 3D surfaces & surface distances are handled in deep mind is different from MONAI: Instead of defining surfaces at surface voxel boundaries, surface points are located on a grid that is shifted by half of the grid spacing along each axis (cf. Fig. 10 in their [paper](https://arxiv.org/abs/1809.04430)):
> 
> grafik
> 
> I think it should be discussed more broadly whether to keep the surface computation currently used in MONAI, or to switch to the deep mind computation, or to establish a switch such that the user can decide on the computation. This choice should be made consistently for all boundary-overlap metrics.

cc @SilviaSeidlitz

### PR DESCRIPTION
**Pull request by wyli**
Fixes #4103

### Description
- considers spacing and subvoxel borders when computing surface Dice
- reimplemented http://medicaldecathlon.com/files/Surface_distance_based_measures.ipynb using pytorch API
- there's possibility of speed up once https://github.com/Project-MONAI/MONAI/issues/1332 is addressed

### Types of changes

- [x] Non-breaking change (fix or new feature that would not break existing functionality).
- [ ] Breaking change (fix or new feature that would cause existing functionality to change).
- [x] New tests added to cover the changes.
- [ ] Integration tests passed locally by running `./runtests.sh -f -u --net --coverage`.
- [x] Quick tests passed locally by running `./runtests.sh --quick --unittests  --disttests`.
- [x] In-line docstrings updated.
- [x] Documentation updated, tested `make html` command in the `docs/` folder.

### CODE REVIEW

File: tests/test_surface_dice.py [hunk 1]
```diff
@@ -330,21 +331,8 @@ def test_asserts(self):
...
 
-        with self.assertRaises(ValueError) as context:
-            SurfaceDiceMetric(class_thresholds=[1, 1], include_background=True)(predictions_no_hot, predictions_hot)
-        self.assertEqual("y_pred and y should be one-hot encoded.", str(context.exception))
-        with self.assertRaises(ValueError) as context:
-            SurfaceDiceMetric(class_thresholds=[1, 1], include_background=True)(predictions_hot, predictions_no_hot)
-        self.assertEqual("y_pred and y should be one-hot encoded.", str(context.exception))
-
         predictions_no_hot = predictions_no_hot.float()
         predictions_no_hot[0, :, 0, 0] = torch.tensor([0.5, 0])
```
Comment (yiheng-wang-nv) [inline tests/test_surface_dice.py:335]: If these tests are remove, `predictions_no_hot` is not used and can also be remove.

File: monai/metrics/utils.py [hunk 2]
```diff
@@ -292,45 +323,410 @@ def prepare_spacing(
...
+        device: device to put the table on.
+
+    Returns:
+        An array of size 256, mapping neighbourhood code to the surface area.
+        ENCODING_KERNEL[3] which is the kernel used to compute the neighbourhood code.
+    """
+    # compute the area for all 256 possible surface elements given a 2x2x2 neighbourhood according to the spacing_mm
+    c = _get_neighbour_code_to_normals_table(device)
+    s = torch.as_tensor(
+        [[[spacing_mm[1] * spacing_mm[2], spacing_mm[0] * spacing_mm[2], spacing_mm[0] * spacing_mm[1]]]],
```
Comment (yiheng-wang-nv) [inline monai/metrics/utils.py:670]: could we also has a simple check for `spacing_mm` ? I think it should has length 3.

File: monai/metrics/utils.py [hunk 3]
```diff
@@ -292,45 +323,410 @@ def prepare_spacing(
...
+
+    Args:
+        spacing_mm: 2-element list-like structure. Pixel spacing along the 1st and 2nd spatial axes.
+        device: device to put the table on.
+
+    Returns:
+        A 16-element array mapping neighbourhood code to the contour length.
+        ENCODING_KERNEL[2] which is the kernel used to compute the neighbourhood code.
+    """
+    first, second = spacing_mm  # spacing along the first and second spatial dimension respectively
```
Comment (yiheng-wang-nv) [inline monai/metrics/utils.py:698]: similar here

**Review by yiheng-wang-nv — COMMENTED** [review]
Thanks @wyli . I checked the changes and this PR looks good to me. Could you address a few  minor issues?

**Comment by wyli** [comment]
> Thanks @wyli . I checked the changes and this PR looks good to me. Could you address a few minor issues?

thanks, the comments are addressed, I also update to use the pytorch APIs as much as possible instead of converting everything into numpy arrays.

**Review by yiheng-wang-nv — APPROVED** [review]
